docs/execution-modes.md

```diff
@@ -141,13 +141,147 @@ def train_model(...):
+
+Let's assume `calculate_average` and `add` are actually doing something
+"heavy" that requires a dedicated container, rather than just performing
+simple arithmetic operations.
+
+People sometimes assume that since `divide`
+is nested inside `calculate_average`, and `divide` is inline, `divide`
+must execute in the same container as `calculate_average`. This is NOT
+correct. This misunderstanding stems from a misunderstanding of how Sematic
+works with Futures. Recall that Sematic functions return futures when you
+call them. That means that `total` in `calculate_average`
+*actually holds a Future instead of a `float`*. So when `divide(total, 3)`
+is called above, the content of `total` is not even known. Therefore how
+could it be executed?
+
+Instead, what happens is this:
+
+1. `add` is called and immediately returns a Future without doing any work
+2. `divide` is called, given the Future from `add` and the constant `3`. It
+also returns immediately without doing any work.
+3. `calculate_average` returns the `Future` output by `divide`
+4. The resolver job analyzes the `Future` coming from `calculate_average` and
+sees that to get the value for it, it must first execute `add` and then
+execute `divide`.
+5. Since `add` is non-inline, the resolver starts a container within which to
+execute `add`. `add` returns an actual `float` which is the sum of `a`, `b`, and
+`c`.
+6. The resolver sees that it now has everything required to execute `divide`, so
+it does so. Since `divide` is inline, the resolver doesn't need to start a new
+container for it, and instead it executes `divide` in its own process.
+
+#### When to use inline?
+
+After walking through the above example, you may be wondering if there's a simple
+way to know when something should be marked as inline vs not. Even if you don't
+follow the above trace of the execution, you can still use `inline` correctly if
+you follow this guidance:
+
+- Any Sematic function doing something "trivial" that executes in a few seconds or
+less and requires negligible CPU or memory should be inline.
+- Any Sematic function which primarily calls other Sematic functions and doesn't
+do any work "of its own" aside from these calls should be inline
+- Any other Sematic functions should NOT be inline. In practice this usually means
+"leaf node" Sematic functions that don't call other Sematic functions and which
+do some "real work."
+
+Most Sematic functions tend to meet the first two criteria, so functions are inline
+by default.
+
+#### Can I make nested functions execute in the parent container?
+
+Let's suppose that you wanted `add` and `divide` to execute in the
+same container as `calculate_average` above. Is that possible? Yes!
+You can just remove the `@sematic.func` decorator from `add` and
+`divide` to make them regular python functions. In this case, they will
+execute just like any other python code--immediately at the time they
+are called, in the same process as the code that called them.
```

Comment:
Maybe remind readers that in this case the function will not be tracked and visualizable in the UI.